### PR TITLE
Add incentive reserved ratio &  Add liquidation threshold

### PR DESCRIPTION
**Pull request by yrong**
@zhoujia6139 @yz89 

After checking the `liquidate_borrow` impl seems for [ED-1392](https://parallelfi.atlassian.net/browse/ED-1392) we do not need to setup another reserve address. Just update liquidator's voucher_balance a little less should be enough and the left part will automatically reserved in our pallet account. Correct me if I'm wrong.

PS: The pr is not ready yet, just ignore the hard coded ratio will remove to runtime config later or as initialized parameter of market.

### CODE REVIEW

**Comment by yrong** [comment]
after some discussion, we add a separated pallet account for the reserved incentive and allow redeem back by governance when required

**Comment by yrong** [comment]
- [x] reserved to separated internal account
- [ ] revamp redeem(all) extrinsic to allow origin from governance
- [ ] fix the breaking test
- [ ] market storage migration
- [ ] benchmark again

**Comment by yrong** [comment]
```
cargo run --bin parallel --features try-runtime -- try-runtime --chain heiko-dev --wasm-execution=compiled on-runtime-upgrade live --uri=wss://heiko-rpc.parallel.fi
    Finished dev [unoptimized + debuginfo] target(s) in 1.24s
     Running `target/debug/parallel try-runtime --chain heiko-dev --wasm-execution=compiled on-runtime-upgrade live '--uri=wss://heiko-rpc.parallel.fi'`
thread 'main' panicked at 'Argument "pallets"
  Arg::is_use_value_delimiter_set is required when Arg::is_require_value_delimiter_set is set.
', /home/ron/.cargo/registry/src/github.com-1ecc6299db9ec823/clap-3.1.3/src/build/debug_asserts.rs:659:5
```

seems related with https://github.com/paritytech/substrate/pull/11017  and fixed in `polkadot-v0.9.18` and later release

@GopherJ  have you met this issue before? are we going to bump parity dependencies in next upgrade window? Meanwhile I guess can only test the storage migration with locally setup.

**Comment by GopherJ** [comment]
@yrong yeah, polkadot-v0.9.19 is released and we are interested in it so I'm making #1632 , it'll be merged tomorrow I think

**Comment by yz89** [comment]
> ```
> cargo run --bin parallel --features try-runtime -- try-runtime --chain heiko-dev --wasm-execution=compiled on-runtime-upgrade live --uri=wss://heiko-rpc.parallel.fi
>     Finished dev [unoptimized + debuginfo] target(s) in 1.24s
>      Running `target/debug/parallel try-runtime --chain heiko-dev --wasm-execution=compiled on-runtime-upgrade live '--uri=wss://heiko-rpc.parallel.fi'`
> thread 'main' panicked at 'Argument "pallets"
>   Arg::is_use_value_delimiter_set is required when Arg::is_require_value_delimiter_set is set.
> ', /home/ron/.cargo/registry/src/github.com-1ecc6299db9ec823/clap-3.1.3/src/build/debug_asserts.rs:659:5
> ```
> 
> during test of [bd288c1](https://github.com/parallel-finance/parallel/commit/bd288c10c80dd4b573c20934caeb1ae089e6d6fc) found the issue and seems related with [paritytech/substrate#11017](https://github.com/paritytech/substrate/pull/11017) and fixed in `polkadot-v0.9.18` and later release
> 
> @GopherJ have you met this issue before? are we going to bump parity dependencies in next upgrade window? Meanwhile I guess can only test the storage migration with locally setup.

Should use release bin rather than debug one.

File: pallets/loans/src/types.rs [hunk 1]
```diff
@@ -47,13 +47,17 @@ pub enum MarketState {
 pub struct Market<Balance> {
     /// The collateral utilization ratio
     pub collateral_factor: Ratio,
+    /// A loose ratio than collateral_factor to avoid liquidate_borrow too casual
+    pub loose_collateral_factor: Ratio,
```
Comment (yz89) [inline pallets/loans/src/types.rs:51]: Let's use liquidation_threshold, it's more clear.

File: pallets/loans/src/lib.rs [hunk 2]
```diff
@@ -476,10 +479,20 @@ pub mod pallet {
                     && market.collateral_factor < Ratio::one(),
                 Error::<T>::InvalidFactor,
             );
+            ensure!(
+                market.loose_collateral_factor < Ratio::one()
+                    && market.loose_collateral_factor > market.collateral_factor,
```
Comment (yz89) [inline pallets/loans/src/lib.rs:484]: I think the liquidation_threshold can be greater than or equal to the collateral factor.
liquidation_threshold >= collateral_factor

File: pallets/loans/src/types.rs (hunk 1 above)
Comment (zhoujia6139) [reply pallets/loans/src/types.rs:51]: Yeah, I'm confused by this loose_collateral_factor.

Comment (yrong) [reply pallets/loans/src/types.rs:51]: is the name `threshold` too vague? I think all current factors like `collateral_factor`, `close_factor` and be some kind of liquidation threshold

**Comment by yrong** [comment]
```
2022-04-21 15:28:44 👜 loans borrow-limit migration passes PRE migrate checks ✅
2022-04-21 15:28:44 migrating loans to Versions::V3
2022-04-21 15:28:44 👜 completed loans migration to Versions::V3
022-04-21 15:28:44 market 100, loose_collateral_factor <wasm:stripped>, liquidate_incentive_reserved_factor <wasm:stripped>
2022-04-21 15:28:44 market 0, loose_collateral_factor <wasm:stripped>, liquidate_incentive_reserved_factor <wasm:stripped>
2022-04-21 15:28:44 market 5002, loose_collateral_factor <wasm:stripped>, liquidate_incentive_reserved_factor <wasm:stripped>
2022-04-21 15:28:44 market 5004, loose_collateral_factor <wasm:stripped>, liquidate_incentive_reserved_factor <wasm:stripped>
2022-04-21 15:28:44 market 1000, loose_collateral_factor <wasm:stripped>, liquidate_incentive_reserved_factor <wasm:stripped>
2022-04-21 15:28:44 market 5003, loose_collateral_factor <wasm:stripped>, liquidate_incentive_reserved_factor <wasm:stripped>
2022-04-21 15:28:44 👜 loans borrow-limit migration passes POST migrate checks ✅
```
migration seems as expected.

**Comment by yrong** [comment]
@GopherJ @MrPai there is another issue found during try-runtime test when enable `LiquidStakingMigrationV3` 

```
2022-04-21 14:53:18 MatchingLedger total_stake_amount: 9101340531076, total_unstake_amount: 0
2022-04-21 14:53:18 MarketCap.get()? 0
2022-04-21 14:53:18 panicked at 'MarketCap storage item not found!', /home/ron/parallel/pallets/liquid-staking/src/migrations.rs:48:9
Error: Input("failed to execute TryRuntime_on_runtime_upgrade: Execution aborted due to trap: wasm trap: wasm `unreachable` instruction executed\nWASM backtrace:\n\n    0: 0x3fffe1 - <unknown>!rust_begin_unwind\n    1: 0x3aca - <unknown>!core::panicking::panic_fmt::h8f11323637b4db3b\n    2: 0x2aee36 - <unknown>!pallet_liquid_staking::migrations::v3::pre_migrate::hc4bfa8a02883da15\n    3: 0x15539b - <unknown>!frame_executive::Executive<System,Block,Context,UnsignedValidator,AllPalletsWithSystem,COnRuntimeUpgrade>::try_runtime_upgrade::h2aa0c3ccb9f5577f\n    4: 0x3a30f - <unknown>!<heiko_runtime::Runtime as frame_try_runtime::runtime_decl_for_TryRuntime::TryRuntime<sp_runtime::generic::block::Block<sp_runtime::generic::header::Header<u32,sp_runtime::traits::BlakeTwo256>,sp_runtime::generic::unchecked_extrinsic::UncheckedExtrinsic<sp_runtime::multiaddress::MultiAddress<<<sp_runtime::MultiSignature as sp_runtime::traits::Verify>::Signer as sp_runtime::traits::IdentifyAccount>::AccountId,()>,heiko_runtime::Call,sp_runtime::MultiSignature,(frame_system::extensions::check_non_zero_sender::CheckNonZeroSender<heiko_runtime::Runtime>,frame_system::extensions::check_spec_version::CheckSpecVersion<heiko_runtime::Runtime>,frame_system::extensions::check_tx_version::CheckTxVersion<heiko_runtime::Runtime>,frame_system::extensions::check_genesis::CheckGenesis<heiko_runtime::Runtime>,frame_system::extensions::check_mortality::CheckMortality<heiko_runtime::Runtime>,frame_system::extensions::check_nonce::CheckNonce<heiko_runtime::Runtime>,frame_system::extensions::check_weight::CheckWeight<heiko_runtime::Runtime>,pallet_transaction_payment::ChargeTransactionPayment<heiko_runtime::Runtime>)>>>>::on_runtime_upgrade::h9afd690cba0f7838\n    5: 0x265001 - <unknown>!TryRuntime_on_runtime_upgrade\n")
```
seems like a bug in migration logic of LS? so temporarily disable in this commit
https://github.com/parallel-finance/parallel/pull/1588/commits/4164b5b6c633716112101031c97fea42eeb37719

File: pallets/loans/src/tests/liquidate_borrow.rs [hunk 3]
```diff
@@ -1,12 +1,14 @@
...
+        new_test_ext, Assets, Loans, MockPriceFeeder, Origin, Ratio, Test, ALICE, BOB,
+        DEFAULT_LIQUIDATE_INCENTIVE_RESERVED_FACTOR, DOT, KSM, USDT,
     },
-    tests::dollar,
+    tests::{dollar, milli_dollar},
     Error, MarketState,
 };
 use frame_support::{assert_noop, assert_ok};
 use primitives::Rate;
+use sp_runtime::traits::Saturating;
```
[the remark this answers is not in the document]
Comment (yrong) [reply pallets/loans/src/tests/liquidate_borrow.rs]: @GopherJ could we reuse same rustfmt as polkadot?
https://github.com/paritytech/polkadot/blob/master/rustfmt.toml

[the remark this answers is not in the document]
Comment (GopherJ) [reply pallets/loans/src/tests/liquidate_borrow.rs]: yeah we can define our rustfmt in a seperate PR I guess

File: pallets/loans/src/lib.rs [hunk 4]
```diff
@@ -1085,6 +1109,37 @@ pub mod pallet {
...
+            asset_id: AssetIdOf<T>,
+            #[pallet::compact] redeem_amount: BalanceOf<T>,
+        ) -> DispatchResultWithPostInfo {
+            T::ReserveOrigin::ensure_origin(origin)?;
+            ensure!(!redeem_amount.is_zero(), Error::<T>::InvalidAmount);
+            let receiver = T::Lookup::lookup(receiver)?;
+            let from = Self::incentive_reward_account_id()?;
+            Self::ensure_active_market(asset_id)?;
+            Self::accrue_interest(asset_id)?;
+            Self::update_earned_stored(&from, asset_id)?;
```
Comment (zhoujia6139) [inline pallets/loans/src/lib.rs:1131]: Should you calculate farming reward for incentive_reward_account? @yz89  if so, then need some change here.

File: pallets/loans/src/migrations.rs [hunk 5]
```diff
@@ -56,25 +63,29 @@ pub mod v2 {
...
-            Markets::<T>::translate::<OldMarket<BalanceOf<T>>, _>(|_key, market| {
+            Markets::<T>::translate::<V2Market<BalanceOf<T>>, _>(|_key, market| {
                 Some(Market {
-                    borrow_cap: 1_000_000_000_000_000u128,
-                    supply_cap: market.cap,
+                    borrow_cap: market.borrow_cap,
+                    supply_cap: market.supply_cap,
                     collateral_factor: market.collateral_factor,
+                    liquidation_threshold: (market.collateral_factor
+                        + market.collateral_factor * DEFAULT_LIQUIDATE_THRESHOLD),
```
Comment (yz89) [inline pallets/loans/src/migrations.rs:75]: Let's add 10% here. 
DEFAULT_LIQUIDATE_THRESHOLD can be defined in migrations.rs, and I think DEFAULT_LIQUIDATION_OFFSET is better.

**Comment by GopherJ** [comment]
the CI issue may be caused by rust toolchain version, can you rebase/merge latest master and try again?

File: pallets/loans/src/lib.rs [hunk 6]
```diff
@@ -1529,14 +1649,27 @@ impl<T: Config> Pallet<T> {
                 Ok(())
             },
         )?;
+        let incentive_reserved_amount = incentive_ratio.mul_floor(collateral_amount);
```
[the remark this answers is not in the document]
Comment (yz89) [reply pallets/loans/src/lib.rs:1652]: No, actually incentive_reserved_amount gets 3.333% which is not what we want.
The formula here should be incentive_reserved_amount = collateral_amount/1.1 * 3%

[the remark this answers is not in the document]
Comment (yz89) [reply pallets/loans/src/lib.rs:1652]: I wonder why this case didn't catch by the unit test.

**Comment by GopherJ** [comment]
weird dont know what happened:)

**Comment by yrong** [comment]
> weird dont know what happened:)

@GopherJ 
test the ci build flow in a forked repo, `make test` is success
https://github.com/yrong/parallel/runs/6147497103?check_suite_focus=true 

seems not related to the version of rust, maybe we can remove the [cache layer](https://github.com/parallel-finance/parallel/blob/64c078b5d797943ac30a9dbe3a83f16cd097f873/.github/workflows/ci.yml#L62-L78) temporarily and try again? Or could we change to [self host runner](https://docs.github.com/en/actions/hosting-your-own-runners/about-self-hosted-runners) , speed of build in github host is a little slow

**Comment by GopherJ** [comment]
> 

yes we may think about changing to self hosted runner

**Review by zhoujia6139 — APPROVED** [review]

**Review by zhoujia6139 — APPROVED** [review]

**Review by GopherJ — APPROVED** [review]

**Pull request merged**